src/main/java/com/streamr/client/dataunion/DataUnion.java

```diff
@@ -0,0 +1,233 @@
+    private DataUnionSidechain sidechain;
+    private Web3j mainnetConnector, sidechainConnector;
+    private Credentials mainnetCred, sidechainCred;
+
+    //use DataUnionClient to instantiate
+    protected DataUnion(DataUnionMainnet mainnet, Web3j mainnetConnector, Credentials mainnetCred, DataUnionSidechain sidechain, Web3j sidechainConnector, Credentials sidechainCred) {
+        this.mainnet = mainnet;
+        this.mainnetConnector = mainnetConnector;
+        this.mainnetCred = mainnetCred;
+        this.sidechain = sidechain;
+        this.sidechainConnector = sidechainConnector;
+        this.sidechainCred = sidechainCred;
+    }
+
+    public boolean waitForDeployment(long pollInterval, long timeout) throws Exception {
+        String code = waitForCodeAtAddress(sidechain.getContractAddress(), sidechainConnector, pollInterval, timeout);
+        return code != null && !code.equals("0x");
+    }
+
+    public String mainnetAddress(){
+        return mainnet.getContractAddress();
+    }
+
+    public String sidechainAddress(){
+        return sidechain.getContractAddress();
+    }
+
+    public boolean isDeployed() throws Exception {
+        //this will check the condition only once:
+        return waitForDeployment(0,0);
+    }
+
+    public void sendTokensToBridge() throws Exception {
+        mainnet.sendTokensToBridge().send();
+    }
+
+    public BigInteger waitForEarningsChange(final BigInteger initialBalance, long pollInterval, long timeout) throws Exception {
+        Web3jUtils.Condition earningsChange = new Web3jUtils.Condition(){
+            @Override
+            public Object check() throws Exception {
+                BigInteger bal = sidechain.totalEarnings().send().getValue();
+                if(!bal.equals(initialBalance)) {
+                    return bal;
+                }
+                return null;
+            }
+        };
+        return (BigInteger) waitForCondition(earningsChange, pollInterval, timeout);
+    }
+
+    public EthereumTransactionReceipt addJoinPartAgents(String ... agents) throws Exception {
+        return new EthereumTransactionReceipt(sidechain.addJoinPartAgents(asDynamicAddressArray(agents)).send());
+    }
+
+    public EthereumTransactionReceipt partMembers(String ... members) throws Exception {
+        return new EthereumTransactionReceipt(sidechain.partMembers(asDynamicAddressArray(members)).send());
+    }
+
+    public EthereumTransactionReceipt joinMembers(String ... members) throws Exception {
+        return new EthereumTransactionReceipt(sidechain.addMembers(asDynamicAddressArray(members)).send());
+    }
+
+
+    /**
+     *
+     * withdraw a member to his own address. Must be that member or admin.
+
+     * @param member
+     * @param amount amout in wei or 0 to withdraw everything
+     * @return
+     * @throws Exception
+     */
+    public EthereumTransactionReceipt withdrawMember(String member, BigInteger amount) throws Exception {
```

Comment:
What this method does, not very clear from the name? Does it withdraw some balance to a member (and does not "withdraw"/"delete" the member)?

Reply:
yes, withdraw balance. same name as contract method and I think JS client

Reply:
I clarified names: withdrawTokensForMember or withdrawTokensForSelfOrAsAdmin